👥 Multiple sessions 👥

nickvergessen opened the pull request:
![spiderman](https://user-images.githubusercontent.com/213943/108230456-1b506a80-7141-11eb-8d7c-dad1b86aeda5.jpg)

* 😿  For now only multiple Talk sessions in different PHP sessions will be supported. So you can be in the same conversation with mobile apps, SIP, browser, etc at the same time, but not in the same chat in 2 tabs of the same browser. The additional burden is currently too much to handle
- [x] Transform the unique attendee key on the sessions table to a normal index
- [x] Make sure that we always select the Talk session of the current PHP session because there could be more than 1
- [ ] In places where the session is not used, the parameter should be set to false/null, to avoid left joining an unneeded database table
- [x] Move integration tests to API v4
- [x] Move UI to API v4
- [x] Remove special handling of lower API versions as the routes don't match anymore
- [x] Add v4 capability and remove others

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -476,7 +477,7 @@ protected function formatRoomV1(Room $room, ?Participant $currentParticipant): a
 		$cleanGuests = false;
 		$participantList = [];
-		$participants = $this->participantService->getParticipantsForRoom($room);
+		$participants = $this->participantService->getParticipantsForRoom($room, true); // FIXME NEEDS the session but can potentially kill APIv1?
```
I think we can just not keep this alive anymore. we can drop this when mobile clients all speak APIv3 by default and just fallback to APIv1 when no APIv3 is provided. that should work for iOS already and Android is on the way.

nickvergessen replied:
this is still the question what to do.
I would introduce a v4 with the current changes that take place to make clear that only Talk sessions of the current PHP session are returned for the current user and basically would like to kill all older room APIs, as their behaviour is inconsistent to how it was before and we can't keep it backwards compatible.

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -1109,7 +1110,7 @@ public function getParticipants(bool $includeStatus = false): DataResponse {
 
 		$maxPingAge = $this->timeFactory->getTime() - 100;
-		$participants = $this->participantService->getParticipantsForRoom($this->room);
+		$participants = $this->participantService->getParticipantsForRoom($this->room, true); // FIXME should return some what session info?
```
This here however is more complicated. We want to show online-state and call icons, but it basically means we would need to select all sessions and then group by actor. We can of course do that, and return a combined call flag of all sessions, but which session id should be returned?
I think currently some clients abuse this for their call screen instead of `CallController::getPeersForCall` (even the web?).

From my POV this should be fixed to use the proper API and then the clients when receiving "please refresh participant list" they refresh participants and peer list for the internal signaling and refresh participants for external signaling + use the data of the event to update their peers.

nickvergessen commented:
Rebase on #5215 for compatibility

nickvergessen commented:
- [ ] Room API returns rooms twice when being 2 times in the conversation
- [ ] UI switches to call view for all sessions instead of only those that joined the call, could be caused by ^ or the RoomController returning data of a random session and the UI just using that participant info.

nickvergessen commented:
Psalm should be happy with https://github.com/nextcloud/server/pull/25816

nickvergessen commented:
- [x] Move integration tests to API v4
- [ ] Move UI to API v4
- [ ] Remove special handling of lower API versions as the routes don't match anymore

nickvergessen commented:
Ready to review. Don't be scared about the size, it's mostly integration tests bumping the required version of the API from none (v1) or v2 or v3 to the new one-and-only v4

PVince81 commented on docs/call.md:
```diff
@@ -1,8 +1,9 @@
+* API v1: 🏁 Removed with API v4
+* API v2: 🏁 Removed with API v4
+* API v3: 🏁 Removed with API v4
+* API v4: Base endpoint `/ocs/v2.php/apps/spreed/api/v4`
```
So this means all clients will stop working with Talk 12 and an upgrade will be mandatory.
If there are no scenarios where someone would not want to upgrade their apps (branding?) I guess this is ok.
Also it won't be possible to first upgrade all client apps first and then the server, so the migration will have to be done "all at once", not gradually, which could mean a bigger downtime for server and clients during that time (but I'm not 100% familiar how upgrades are done in enterprise)

Probably we'll need to make this very clear in release notes and communications.
"Don't upgrade to NC 22 if you can't upgrade all the client apps also at the same time".

Just thinking out loud. I understand that there is no way to keep compatibility.

nickvergessen replied:
> Also it won't be possible to first upgrade all client apps first and then the server, so the migration will have to be done "all at once", not gradually, which could mean a bigger downtime for server and clients during that time (but I'm not 100% familiar how upgrades are done in enterprise)

No in theory apps can already start rolling out, the capability will tell them if v4 is available on the server and otherwise they use the old v1-3 apis. It's just not possible anymore to serve v1-3 from a current server. but clients can be compatible with v1-4 at the same time (with some if v4 then X else Y logic)

PVince81 commented on lib/Room.php:
```diff
@@ -402,28 +403,47 @@ public function getPropertiesForSignaling(string $userId, bool $roomModified = t
+					$query->expr()->eq('a.id', 's.attendee_id')
+				));
+			} else {
+				$helper->selectSessionsTable($query); // FIXME PROBLEM
```
fixme problem?

nickvergessen replied:
Fixed

PVince81 commented on src/services/participantsService.js:
```diff
@@ -158,7 +158,7 @@ const leaveConversation = async function(token) {
  */
 const leaveConversationSync = function(token) {
-	axios.delete(generateOcsUrl('apps/spreed/api/v3/room', 2) + token + '/participants/active')
+	axios.delete(generateOcsUrl('apps/spreed/api/v4', 2) + `room/${token}/participants/active`)
```
this would not escape the token arg, but it's ok

sadly generateOcsUrl doesn't seem to accept params, but generateUrl does: https://github.com/nextcloud/nextcloud-router/blob/master/lib/index.ts#L53

I've raised it here for later: https://github.com/nextcloud/nextcloud-router/issues/194

since the token is known to be safe, ok to keep

nickvergessen replied:
Oh, I didn't know this as all APIs I use are OCS. sounds like a good enhancement

PVince81 replied:
Yeah, I think it used to be that way when it was still part of server/core and didn't get ported that way.

PVince81 commented on src/store/participantsStore.js:
```diff
@@ -114,6 +125,32 @@ const mutations = {
+			if (!state.connecting[token]) {
+				Vue.set(state.connecting, token, {})
+			}
+			Vue.set(state.connecting[token], sessionId, flags)
```
so we're setting the exact same values/structure to "connecting" like "inCall" ?
either redundant or a copy-paste mistake ?

since none of these have any other getters than `isInCall` and `isConnected` the data structure could be merged into a single one but still keep both getters ?

nickvergessen replied:
I find it easier with this explicit way instead of wrapping my head around existing and non-existing attributes and also we need to block "faulty" updates from the participant list update etc. But feel free to do it in a follow up PR when you think it's worth it

PVince81 replied:
Not sure if you understood what I meant.

It seems that both `isInCall` and `isConnected` always contain the exact same data "flags".
There is no scenario I could find where they would contain different data, so why use two attributes instead of just one ?

PVince81 replied:
the attribute could be called `callFlags[token]` since it contains the flags.

PVince81 replied:
ah! okay, found it: so `finishedConnecting` is then deleting the token entry after success. that's the bit I missed.

in that case maybe just store a boolean in `connecting` instead of all the flags to avoid confusion

PVince81 replied:
I'm fine leaving everything as is. it's just likely that the next dev coming over here might wonder as well

PVince81 commented on src/store/participantsStore.js:
```diff
@@ -247,23 +284,48 @@ const actions = {
 
 	async joinCall({ commit, getters }, { token, participantIdentifier, flags }) {
+		if (!participantIdentifier?.sessionId) {
+			console.error('Trying to join call without sessionId')
```
no return, so we still want to attempt the server call?

nickvergessen replied:
right, fixed

PVince81 commented on tests/integration/features/callapi/public.feature:
```diff
@@ -5,105 +5,105 @@ Feature: callapi/public
+    Then user "guest" leaves room "room" with 200 (v4)
+    Then user "participant1" sees 1 peers in call "room" with 200 (v4)
+    And user "guest" sees 0 peers in call "room" with 404 (v4)
+    Then user "participant1" leaves call "room" with 200 (v4)
```
after scrolling through pages and pages if sprinkled `(v4)` I wonder if we even need to mention it in the tests, considering that v4 is the new default and we're not going to test versions < v4 and only when v5 comes in, if incompatible, we would require to specify it

nickvergessen replied:
So the thing is, the version was optional until now. I force it to now, because I noticed that all the tests used the v1 in 99% of the cases, so they didn't use/test the methods we use in the UI etc.
Sure we could go back to optional and just use v4 as fallback, but I'd rather be explicit making everyone aware that any other version is not tested.

PVince81 replied:
agreed.

when scrolling further I saw we still had some v1 steps, so making it explicit will make sure there is no confusion

PVince81 reviewed (commented):
Some minor comments, especially some duplication in frontend.

Haven't tested yet

PVince81 commented:
> In places where the session is not used, the parameter should be set to false/null, to avoid left joining an unneeded database table

See checkbox, is it done ? I saw a lot of places where you set it to false alreay.

PVince81 commented:
Tests:
- [x] TEST: could join twice with same user
- [x] TEST: raising and lowering hand from both sessions properly updates the indicator in participant list for others (aggregated)
- [x] TEST: kicking out user from conversation terminated both sessions
- [ ] TEST: joining twice as the same guest
- [ ] TEST: files sidebar chat
- [ ] TEST: video verification (regresison check)
- [ ] TEST: joining+call twice as self in 1-1 conversation

nickvergessen commented on lib/Events/RemoveParticipantEvent.php:
```diff
@@ -31,14 +32,26 @@ class RemoveParticipantEvent extends ParticipantEvent {
 								Participant $participant,
-								string $reason) {
+								string $reason,
+								array $sessions = []) {
```
Since i moved the listener to the before event this modification might not be neededd

PVince81 commented:
@nickvergessen seems the multi-session thing is not working with HPB (receiving duplicate session error), only internal signaling works. I didn't see any statement about this so not sure if it's a bug or whether my version of the HPB is outdated.

nickvergessen commented:
I didn't check HPB yet, it also needs some adjusting as it calls v3 api itself. I notified Jojo already about the changes and will test that part more deeply monday

PVince81 commented:
- [ ] BUG: joining as the same guest twice in the same browser (not technically the same session) works for chatting but the call part is buggy

Not sure if we intended to support this. If not we should catch it and return a "duplicate session" error like for regular accounts.

nickvergessen commented:
> BUG: with HPB joining as the same guest twice in the same browser (not technically the same session) works for chatting but the call part is buggy

Same browser is the same php session and therefor not something we can fix atm

PVince81 commented:
I think I'm done testing. Apart from the HPB case and some corner issues, it seemed to work fine :smile:

See what I tested here https://github.com/nextcloud/spreed/pull/5194#issuecomment-791420898
Will need to redo this once it works with HPB.

nickvergessen commented:
> * TEST: could join and call twice with same user
>    * HPB no_entry_sign getting duplicate session

So this is an easy fix. The guest thing however is impossible atm. Also for the HPB users break as they can join the same conversation twice with 2 different session ids in the same browser. The problem is we don't have access to the PHP session when the signaling server tells us "Hey, user X joined with session id Y" and so we always create a new session. The problem is one of them has a session that doesn't match the session id in the PHP session and therefor is not match/returned correctly.

nickvergessen commented:
> system messages for join/leave only displayed once when relevant 

So the question of course is what "relevant" is. I would just display it all the time for calls as it might be important to know that another device disconnected.

nickvergessen commented:
Added 3 commits and rebased.
Works as expected now with the HPB from my POV.

Only thing is guests see "Conversation not found" after their previous session gets kicked because of duplication.

PVince81 commented:
I've tested with HPB and all seemed to work fine.

There's only a minor issue with raise hand because we don't aggregate HPB style session id in the participant list, but perhaps we can fix this separately.

PVince81 reviewed (approved):
:+1:

PVince81 commented:
some integration tests are failing still: https://drone.nextcloud.com/nextcloud/spreed/1816/4/3
```
/drone/server/apps/spreed/tests/integration/features/conversation/one-to-one.feature:121
    /drone/server/apps/spreed/tests/integration/features/conversation/one-to-one.feature:130
    /drone/server/apps/spreed/tests/integration/features/conversation/remove-participant.feature:10
    /drone/server/apps/spreed/tests/integration/features/conversation/remove-participant.feature:18
    /drone/server/apps/spreed/tests/integration/features/conversation/remove-participant.feature:30
    /drone/server/apps/spreed/tests/integration/features/conversation/remove-participant.feature:67
    /drone/server/apps/spreed/tests/integration/features/conversation/remove-participant.feature:79
    /drone/server/apps/spreed/tests/integration/features/conversation/remove-participant.feature:94
```

nickvergessen commented:
Ah yeah... that is what the fixme was about 😅

PVince81 commented:
Also, I still have some features in my mind like "[ask to unmute](https://github.com/nextcloud/spreed/issues/4566)" or "[mute from sidebar](https://github.com/nextcloud/spreed/issues/4412)". Now since we only show a participant once during a call there the UX for such would be more tricky as we'd need to either ask the user which session they want to mute (maybe only one is unmuted by mistake) or do it for all of them.

Also not sure if it's a good idea to temporarily show one participant as multiple entries in the participant list during a call. Or group them somehow with an expandable arrow. Feels like it's going too far.

nickvergessen commented:
> Also, I still have some features in my mind like "ask to unmute" or "mute from sidebar". Now since we only show a participant once during a call there the UX for such would be more tricky as we'd need to either ask the user which session they want to mute (maybe only one is unmuted by mistake) or do it for all of them.

Yeah, let's discuss this separatly. For now muting works on the call view and its okay for now.

nickvergessen commented:
> raising and lowering hand from both sessions properly updates the indicator in participant list for others (aggregated) no_entry_sign the participant list seems tied only to the first session so the hand only appears when first session raises/lowered, might need some tweaks for session id matching / aggregation => raise for separate fixing ?

https://github.com/nextcloud/spreed/issues/5349

> Only thing is guests see "Conversation not found" after their previous session gets kicked because of duplication, but lets fix that later.

https://github.com/nextcloud/spreed/issues/5350

PVince81 commented:
:tada:
